### PR TITLE
[Feature,Doc] QValue refactoring and QNet + RNN tuto

### PR DESCRIPTION
**Pull request by vmoens**
- init
- amend

## Description

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by vmoens** [comment]
Note: there is still some magic that needs to be done when we pass 2 consecutive trajectories to the RNN, but it can be handled without too much trouble (I think)
@matteobettini I will do that with padding. Don't tell me, I know!

File: torchrl/modules/tensordict_module/exploration.py [hunk 1]
```diff
@@ -447,17 +447,19 @@ def step(self, frames: int = 1) -> None:
...
                 warnings.warn(
                     f"The tensordict passed to {self.__class__.__name__} appears to be "
-                    f"missing the 'step_count' entry. This entry is used to "
+                    f"missing the 'is_init' entry. This entry is used to "
                     f"reset the noise at the beginning of a trajectory, without it "
                     f"the behaviour of this exploration method is undefined. "
                     f"This is allowed for BC compatibility purposes but it will be deprecated soon! "
-                    f"To create a 'step_count' entry, simply append a StepCounter "
-                    f"transform to your environment with `env = TransformedEnv(env, StepCounter())`."
+                    f"To create a 'step_count' entry, simply append an torchrl.envs.InitTracker "
```
Comment (matteobettini) [inline torchrl/modules/tensordict_module/exploration.py:457]: Replace step_count with is_init?

File: torchrl/modules/tensordict_module/rnn.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        lstm (torch.nn.Module): a :class:`torch.nn.LSTM` instance. For simplicity, this class does not
+            handle the construction of the module.
+        in_keys (list of str): a triplet of strings corresponding to the input value,
+            first and second hidden key.
+        out_keys (list of str): a triplet of strings corresponding to the output value,
+            first and second hidden key.
+            .. note::
+              For a better integration with TorchRL's environments, the best naming
+              for the output hidden key is ``("next", "hidden0")`` or similar, such
+              that the hidden values are passed from step to step during a rollout.
```
Comment (matteobettini) [inline torchrl/modules/tensordict_module/rnn.py:47]: Why are we asking the keys like this?

In my mind i would have 3 key arguments:
- in_keys # in keys apart from hidden state (not constrained to 1)
- out_keys # out keys apart from hidden state (not constrained to 1)
- hidden_keys = (hidden0, hidden1) # this is by default like this and the init automatically adds these as input keys and the version of these with preappended next to the output keys

File: torchrl/modules/tensordict_module/rnn.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    Examples:
+        >>> from torchrl.envs import TransformedEnv, InitTracker
+        >>> from torchrl.envs.libs.gym import GymEnv
+        >>> from torchrl.modules import MLP
+        >>> from torch import nn
+        >>> from tensordict.nn import TensorDictSequential as Seq, TensorDictModule as Mod
+        >>> env = TransformedEnv(GymEnv("Pendulum-v1"), InitTracker())
+        >>> lstm = nn.LSTM(input_size=env.observation_spec["observation"].shape[-1], hidden_size=64, batch_first=True)
+        >>> lstm_module = LSTMModule(lstm, in_keys=["observation", "hidden0", "hidden1"], out_keys=["intermediate", ("next", "hidden0"), ("next", "hidden1")])
```
Comment (matteobettini) [inline torchrl/modules/tensordict_module/rnn.py:64]: With the comment above this becomes
`LSTMModule(lstm, in_keys=["observation"], out_keys=["intermediate"])`

File: torchrl/modules/tensordict_module/rnn.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            fields={
+                action: Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.float32, is_shared=False),
+                done: Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.bool, is_shared=False),
+                intermediate: Tensor(shape=torch.Size([64]), device=cpu, dtype=torch.float32, is_shared=False),
+                is_init: Tensor(shape=torch.Size([1]), device=cpu, dtype=torch.bool, is_shared=False),
+                next: TensorDict(
+                    fields={
+                        hidden0: Tensor(shape=torch.Size([1, 64]), device=cpu, dtype=torch.float32, is_shared=False),
+                        hidden1: Tensor(shape=torch.Size([1, 64]), device=cpu, dtype=torch.float32, is_shared=False)},
+                    batch_size=torch.Size([]),
```
Comment (matteobettini) [inline torchrl/modules/tensordict_module/rnn.py:78]: Isn't it counterintuitive that 
obs -> intermediate -> action are all at root level but the hidden state is in next?

What again is the reason for this?

**Review by matteobettini — COMMENTED** [review]
Looks good, I left some questions!

- I like that the 2 hidden states are kept separate and I think that is the best choice
- is_init is used nicely in my opinion. We need a flag that needs when to reset and it does exactly that.

**Comment by matteobettini** [comment]
> EDIT: Problem solved, feature implemented!!

We can't escape from padding in rnns right?

**Comment by vmoens** [comment]
> We can't escape from padding in rnns right?

There is the packed_sequence artifact but i'm not entirely sure that it solves the problem.
My implementation of padding / unpadding is 2x faster than other torch tools btw

File: torchrl/modules/tensordict_module/rnn.py (hunk 4 above)
Comment (vmoens) [reply torchrl/modules/tensordict_module/rnn.py:78]: the hidden is not used at time t but t+1 (obs, intermediate and action are for t and t only). It's probably a bit cleaner to have it in next as this is where it belongs... 

so you pack hidden in next such that you find it back when calling step_mdp
Per se we could avoid that, since I think most calls to step_mdp have keep_other=True (it is the default, it is the case for env.rollout and it is the case for the collectors!).
But if someone by mistake calls step_mdp with keep_other=False she'll loose the hidden, and since we populate the hidden with 0s when not found no error will be raised (outch!). 

So i'd rather avoid that scenario as much as I can!

Now i'm not against the idea of having it out of "next" if you find it's clearer.

Comment (matteobettini) [reply torchrl/modules/tensordict_module/rnn.py:78]: I see. So we put it in next because we know for sure that a step_mdp will happen.

Comment (vmoens) [reply torchrl/modules/tensordict_module/rnn.py:78]: in the algo yes but it's your choice...

**Comment by smorad** [comment]
Awesome, really looking forward to this!

I think it makes sense to write the recurrent state to `next`. It just might make things tricky with when/who is calling `step_mdp`. Are there any footguns here, e.g. collector calling `step_mdp` and reversing `state, (next, state)` before placing them into the replay buffer?

Would it make sense to use `td[(prev, done)]` instead of `is_init`? Perhaps there are some edge cases I haven't considered, or maybe this is tricky to implement.

Opinion: I think `set_temporal_mode` could be a little confusing. Is this different than inference (collector) mode, or just a synonym for inference? If they are indeed the same, maybe it makes sense to have `if self.eval: ...` in the module forward pass instead of an explicit time-batch/single-step mode.

I would really prefer that `packed_sequence` is not used here. It's brought me nothing but pain.

File: torchrl/modules/tensordict_module/rnn.py (hunk 4 above)
Comment (smorad) [reply torchrl/modules/tensordict_module/rnn.py:91]: I have to double-check, but I think RNNs are usually written as

`y_t, s_t = rnn(x_t, s_{t-1})`

in books/papers. I agree that perhaps `prev_state, state` makes more sense than `state, next_state`.

Comment (vmoens) [reply torchrl/modules/tensordict_module/rnn.py:91]: There's what papers say and how it aligns with the rest of the API.
ie. I'm not sure how we could mark the input as being "prev" and the current as being "current" if the only containers we have are "current" and "next". Maybe a proper doc that mirrors what you're saying here?

**Comment by vmoens** [comment]
> Awesome, really looking forward to this!
> 
> I think it makes sense to write the recurrent state to `next`. It just might make things tricky with when/who is calling `step_mdp`. Are there any footguns here, e.g. collector calling `step_mdp` and reversing `state, (next, state)` before placing them into the replay buffer?

If you want to store every `h_{t-1}, s_t, a_t, h_t, s_{t+1}` following the nomenclature above, this will correspond to

```
TensorDict(
    fields={
        hidden: Tensor(shape=torch.Size([4, 16]), device=cpu, dtype=torch.int64, is_shared=False),
        action: Tensor(shape=torch.Size([4, 1]), device=cpu, dtype=torch.int64, is_shared=False),
        observation: Tensor(shape=torch.Size([4, 2]), device=cpu, dtype=torch.bool, is_shared=False),
        next: TensorDict(
            fields={
                hidden: Tensor(shape=torch.Size([4, 16]), device=cpu, dtype=torch.int64, is_shared=False),
                observation: Tensor(shape=torch.Size([4, 2]), device=cpu, dtype=torch.bool, is_shared=False),
            batch_size=torch.Size([4]),
            device=cpu,
            is_shared=False),
    batch_size=torch.Size([4]),
    device=cpu,
    is_shared=False)
``` 

> Opinion: I think `set_temporal_mode` could be a little confusing. Is this different than inference (collector) mode, or just a synonym for inference? If they are indeed the same, maybe it makes sense to have `if self.training: ...` in the module forward pass instead of an explicit time-batch/single-step mode.

I agree it's not ideal.
First, we had a similar discussion [here](https://github.com/pytorch/rl/issues/875) if your want to have a look.

First, let's put the cards on the table. There are various things that are related but not exactly identical:
- Train / eval mode
- Explorative / exploitative mode
- temporal / static mode (eg example here)
- grad on / off mode (eg example here)
When doing inference and collecting data for training you will be in 
  eval + explorative + static + grad off
When testing the policy in "real" setting you will be 
  eval + exploitative + static + grad off
When training the policy you will be in
  train + exploitative* + temporal** + grad on***

* not always true, sometimes you want to sample the action, sometimes not
** not always true: in dreamer you simulate trajectories during training that are fully reparametrized, and you're in static mode bc the RNN feeds itself with the action that results from the previous call.
*** depending on the loss (policy / critic / model in MBRL) you may want to keep some ops out of the graph or just detach some variables (not talking about meta-RL!)

Gradient and exploration are currently handled via "exploration_type" and the torch "grad_mode" environment variables.
Train and eval are handled via the module attribute. I hesitate for temporal, but I can see both cases where you may want to use static / temporal at both training and inference time. I thought that giving the user full power over what happens internally was the easiest option to avoid any kind of confusion.

If it's about naming i'm happy to reconsider the name oc. I'm actually looking forward to have more input on this from the community: what's the best way of approaching all of these "modes" that -- to me -- seem very algorithm-dependant. 
Just think about SARSA vs QMax, where the action selection is completely different. We can easily do that via "exploration_type", but it's neither related to eval/train, gradient off/on or else.

> I would really prefer that `packed_sequence` is not used here. It's brought me nothing but pain.

Got it, it's a headache on our side too to let's just not use it.

**Comment by smorad** [comment]
That tensordict format looks good to me.

Yeah that's a good point with the modes. In that case it's better to err on the side of flexibility than simplicity. I think perhaps a small table in the docs mapping each of these modes to a 1-2 sentence description would help clear up any ambiguities. I do like the way exploration etc. is currently handled. Is `set_recurrent_mode` more specific than `set_temporal_mode`? Could either be `batch_time` or `sequential_time`.

nit: `hidden/state` are super overloaded terms. I think we should be explicit with `recurrent_state_h, recurrent_state_c` or `rstate_h, rstate_c`, which will help new users when they are grepping through the codebase.
